addons/website_mail/controllers/main.py

```diff
@@ -22,6 +24,17 @@ def website_message_subscribe(self, id=0, object=None, message_is_follower="on",
         if request.env.user != request.website.user_id:
             partner_ids = request.env.user.partner_id.ids
         else:
+            error_captcha = None
+            try:
+                if not request.env['ir.http']._verify_request_recaptcha_token('website_mail'):
+                    error_captcha = _("Suspicious activity detected by Google reCaptcha.")
+
+            except (ValidationError, UserError) as e:
+                error_captcha = e.args[0]
+            if error_captcha:
+                return json.dumps({
+                    'error': error_captcha,
+                })
```

Comment:
Why not verify recaptcha at the very beginning of the method?
See [this other json route you mentioned](https://github.com/odoo/odoo/blob/8873fe7240e505553650fc15aea1ea3a299acc0f/addons/website_mass_mailing/controllers/main.py#L28-L32)